pkg/image/apis/image/validation/whitelist/whitelister_test.go

```diff
@@ -0,0 +1,323 @@
+
+	"k8s.io/apimachinery/pkg/util/diff"
+
+	serverapi "github.com/openshift/origin/pkg/cmd/server/api"
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+)
+
+func mkAllowed(insecure bool, regs ...string) serverapi.AllowedRegistries {
+	ret := make(serverapi.AllowedRegistries, 0, len(regs))
+	for _, reg := range regs {
+		ret = append(ret, serverapi.RegistryLocation{DomainName: reg, Insecure: insecure})
+	}
+	return ret
+}
+
+func TestRegistryWhitelister(t *testing.T) {
+	for _, tc := range []struct {
+		name      string
+		transport WhitelistTransport
+		whitelist serverapi.AllowedRegistries
+		hostnames map[string]error
+		difs      map[imageapi.DockerImageReference]error
+		pullSpecs map[string]error
+	}{
+		{
+			name:      "empty whitelist",
+			transport: WhitelistTransportSecure,
+			whitelist: mkAllowed(true),
+			hostnames: map[string]error{
+				"example.com":    fmt.Errorf(`registry "example.com" not allowed by whitelist {  }`),
+				"localhost:5000": fmt.Errorf(`registry "localhost:5000" not allowed by whitelist {  }`),
+			},
+			difs: map[imageapi.DockerImageReference]error{
+				{Registry: "docker.io", Namespace: "library", Name: "busybox"}: fmt.Errorf(`registry "docker.io" not allowed by whitelist {  }`),
+				{Name: "busybox"}:                                              fmt.Errorf(`registry "docker.io:443" not allowed by whitelist {  }`),
+			},
+		},
+
+		{
+			name:      "allow any host with secure transport",
+			transport: WhitelistTransportSecure,
+			whitelist: mkAllowed(false, "*"),
+			hostnames: map[string]error{
+				"docker.io":      nil,
+				"example.com":    nil,
+				"localhost:443":  nil,
+				"localhost:5000": fmt.Errorf(`registry "localhost:5000" not allowed by whitelist { "*:443" }`),
+				"localhost:80":   fmt.Errorf(`registry "localhost:80" not allowed by whitelist { "*:443" }`),
+				"localhost":      nil,
+			},
+			difs: map[imageapi.DockerImageReference]error{
+				{Registry: "docker.io", Namespace: "library", Name: "busybox"}: nil,
+			},
+		},
+
+		{
+			name:      "allow any host with insecure transport",
+			transport: WhitelistTransportInsecure,
+			whitelist: mkAllowed(true, "*"),
+			hostnames: map[string]error{
+				"docker.io":      nil,
+				"example.com":    nil,
+				"localhost:443":  fmt.Errorf(`registry "localhost:443" not allowed by whitelist { "*:80" }`),
+				"localhost:5000": fmt.Errorf(`registry "localhost:5000" not allowed by whitelist { "*:80" }`),
+				"localhost:80":   nil,
+				"localhost":      nil,
+			},
+			difs: map[imageapi.DockerImageReference]error{
+				{Registry: "docker.io", Namespace: "library", Name: "busybox"}: nil,
+			},
+		},
+
+		{
+			name:      "allow any host with any transport",
+			transport: WhitelistTransportAny,
+			whitelist: mkAllowed(true, "*"),
+			hostnames: map[string]error{
+				"docker.io":      nil,
+				"example.com":    nil,
+				"localhost:443":  fmt.Errorf(`registry "localhost:443" not allowed by whitelist { "*:80" }`),
```

Comment:
why isn't 443 not allowed for "any transport"?  especially when :80 is allowed?

Reply:
`mkAllowed` creates `AllowedRegistries` with `Insecure` flag set. Therefor the whitelist looks like the error suggests: `{ "*:80" }`. Thus any given hostname must either have `:80` port specified or be passed with one of `{any, insecure}` transport.

Reply:
ok i think i understand this now as well.